### DISCUSSION TITLE
Programmatically draw a rectangle into the annotationLayer / canvas

### DISCUSSION THREAD

**Discussion by jcurtisKSM** [stephanrauh/ngx-extended-pdf-viewer#2132] [Q&A] [answered]
I have a requirement to add a semi-transparent rectangle overlay to the PDF.  I have inspected the use of the Text Layer from what the component provides, unfortunately I have a mixture of IRS forms and text, and the simple <span> search with highlight will not work.

I have four-corner box points to draw a box, which I can do with regular canvas.  What I would like to do is have access to build an object as if I used the "draw" editor, but access the internal canvas object created to draw a rectangle on that.  Lastly, being a "known" object to pdf viewer, be able to programmatically delete that object as if you clicked the "delete / trash can" icon.

**Comment by stephanrauh** [comment]
@juqing27 Can you chime in, please? Your approach using https://github.com/highkite/pdfAnnotate sounds promising. I believe that library can solve the problem James describe. Can you tell us about your experience?

**Reply by juqing27** [reply]
James and Stephan - glad to join the discussion. A few notes from me - 

1) in [this thread](https://github.com/stephanrauh/ngx-extended-pdf-viewer/issues/2090#issuecomment-1915422629), Stephan pointed to me that we can access pdf.js API from the global object `PDFViewerApplication`. From then on I was able to use the pdfAnnotation library together with ngx-extended-pdf-viewer. After installing both libraries (ngx-extended-pdf-viewer and pdfAnnotate), initiate the pdfAnnotate function like this: 

```
PDFViewerApplication.pdfDocument.getData().then((data: Uint8Array) => {
          this.annotationFactory = new AnnotationFactory(data);})
```
and then you can use the annotationFactory like this:
```
this.annotationFactory.createSquareAnnotation(
    {
    page: pageN-1
    rect: [scaledLeft, adjustedTop, scaledLeft + scaledWidth, adjustedTop + scaledHeight],
    contents: annotContent,
    author: coord,
    color: color,
    fill: {r:100, g:100, b:0},
    opacity: 0.5,
    subject: "a note",}
  );  
```
I can share more code if the above is not clear enough. I haven't tried but they also have deleteAnnotation() and getAnnotation(), which should help you find an annotation and delete/manipulate it.

2) Caveat: It seems that the highKite pdfAnnotate library is not being maintained anymore. There are also a few types of annotations that they planned to integrate but didn't, like Stamp and Caret. What they already have (square, circle, etc.) is good enough for my use though. 

3) One challenge in my use case is to precisely locate the search results on pdf page so annotations can be placed accordingly. I'll open a new discussion for that so I don't sidetrack this one too much. I'll link that one here if you are interested/need to solve similar problem. (to clarify this is NOT due to bug or limitation of the pdfAnnotate library. It is doing its job - putting annotations where it's told to. Just generally hard to perfectly locate a pdf element and convert the coordinates in web to coordinates on pdf page, etc.)

4) Alternative to the pdfAnnotate library - I wanted to look at the iText java library too and haven't. That seems to have annotation function as well. Maybe someone in the community already did...?

**Reply by stephanrauh** [reply]
- iText is a great library, but it's Java, so it doesn't integrate as seamlessly as pdfAnnotate. If you're ready to do the extra server round-trip, I can recommend iText. Just make sure you read the license texts. A couple of years ago, they ran into problems which forced them to go commercial. AFAIK the free version is back, but there's also a commercial version that's worth evaluating.
- Thanks for pointing out that pdfAnnotate has (probably) been abandoned. That's really a problem with open source projects! Everybody's using them, but few people keep them alive. Sigh!
- Just a short side remark on your `Promise.then()` code: `async...await` is a real life-saver. Many convoluted algorithms become simple with async-await. Check out my article: https://www.beyondjava.net/angular-async-await

**Reply by juqing27** [reply]
Thanks for the advice @stephanrauh ! I'm new to Angular and JS and am still learning. I'll take a look at your article.

@jcurtisKSM I started another discussion regarding locating elements on PDF page #2134 . If you've got some idea that's very appreciated and please advise :) Otherwise, it has more of my code regarding using pdfAnnotate with ngx-extended-pdf-viewer. Hope that's somewhat helpful for you. 

Qing

**Reply by jcurtisKSM** [reply]
@juqing27 and @stephanrauh - thank you for the replies.  I will get to work on this and see what I come up with.  The idea for the programmatic approach is to have the pdf viewer next to a list of form fields.  When the form field has focus, I use bounding regions to draw a rectangle of where the form field (and page) exist in the pdf.  

When another form field is focused, I remove the previous rectangle and generate the new one.  It was apparent that if I could create this within the confines of the control, I could take advantage of the rectangle auto scaling with resizing, etc.

**Reply by stephanrauh** [reply]
Clever idea! There's a different approach you can try. If the PDF is editable, the form fields of the PDF are actual `<input>` tags. So you can use CSS to add a colorful border around these fields. The field `id` and `name` attributes often have weird values, but in most cases, you can define a CSS selector that returns the correct `<input>` field by variable name.
<img width="1369" alt="image" src="https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/fd947185-93ef-4684-872f-b4cb01119f61">

**Reply by jcurtisKSM** [reply]
I like that idea too.  Unfortunately, for us, these PDFs could be generated (aka "printed" to PDF) or scanned in, and not have the fillable form embedded in them.  

I do have bounding region coordinates from an OCR that is higher in the process where I "know" where the fields exist on a page, and I can write them directly on the canvas but the problems are:
1.  The rectangle needs to scale upon resizing of the pdf (PDF Viewer normal controls used for resizing)
2.  The rectangle needs to be removed if another field is focused in the adjacent form  (could work if we just "cleared all rectangles")
3.  Being that I'm writing directly to the canvas, ctx.clearRect(...) will blank out the pdf page of the canvas. 

I feel I'm really close.  For example, here are my two Typescript functions that I'm using to create the rectangles plus do the conversion from inches to pixels for my bounding regions.

```ts
// =======================================
drawRectangleOnMainCanvas(clear:boolean = false): void {
    if (this.boundingRegion && this.boundingRegion.trim() !== '') {
      try {
        // Parse the string as JSON
        const data = JSON.parse(this.boundingRegion);
        if (data && data.length > 0) {
          const points = data[0][0].polygon; // Take the first polygon points
    
          const canvasWidth = 507;
          const canvasHeight = 657;
    
          // Make sure canvas element is selected after it is available in the DOM
          const canvas = document.querySelector('canvas') as HTMLCanvasElement;
    
          if (canvas) {
            this.drawSemiTransparentRectangle(canvas, points, canvasWidth, canvasHeight, clear);
          }
        }    
      } catch (error) {
        // Log the error and/or handle it as needed
        console.error('Failed to parse bounding region data or was empty:', error);
      }
    }
    
}

drawSemiTransparentRectangle(canvas: HTMLCanvasElement, points: Array<{ x: number, y: number }>, pageWidth: number, pageHeight: number, clear: boolean = false): void {
  if (!canvas.getContext) {
      return;
  }
  
  const ctx = canvas.getContext('2d');
  if (!ctx) {
      return;
  }

  const dpi = 96; // Standard screen DPI

    // Convert inches to pixels
    const xs = points.map(p => p.x * dpi);
    const ys = points.map(p => p.y * dpi);
    
    let minX = Math.min(...xs);
    let minY = Math.min(...ys);
    let maxX = Math.max(...xs);
    let maxY = Math.max(...ys);

    // Adjust for negative width and height by setting starting point to top-left corner
    const width = maxX - minX;
    const height = maxY - minY;
    if (width < 0) {
        minX += width;
    }
    if (height < 0) {
        minY += height;
    }

    //ctx.clearRect(0, 0, canvas.width, canvas.height); // Clear previous drawing
    ctx.scale(this.currentPdfScale, this.currentPdfScale); // Apply the scale
    if (clear) {
      //ctx.fillStyle = 'rgba(0, 0, 255, 0.0)';
      ctx.clearRect(0, 0, canvas.width, canvas.height); // Clear previous drawing
    } else {
      ctx.fillStyle = 'rgba(0, 0, 255, 0.5)';
    }
    
    ctx.fillRect(minX, minY, Math.abs(width), Math.abs(height)); // Use absolute values for width and height
}
// =======================================
```
Additionally, I tried to install the annotatepdf into my angular project and it was complaining about 'fs' being used by annotatepdf.  Apparently, isn't not happy about that, so I have abandoned that route for the time being.

**Reply by stephanrauh** [reply]
I see. Ah, you're drawing on the actual canvas? I wasn't aware of this. I still thought about annotatepdf. That should be possible if you're drawing the rectangles after catching the event `(pageRendered)` and possibly `(currentZoomFactor)` or `(zoomChange)`.

**Reply by jcurtisKSM** [reply]
Just to add to the complexity (glutton for punishment), the viewer exists in a parent angular component, and I have form fields that are in the parent component, and two child components (lots of repeating fields for the child components).

This is my declaration of the viewer plus an example of an "onfocus" event for a text field in a child component.

ParentComponent.html
```html
<div class="pdf-viewer-container">
      <ngx-extended-pdf-viewer 
      [src]="PdfUrl" 
      [(page)]="PdfPageNumber"
      pageViewMode="multiple"
      [zoom]="zoomSetting"
      (zoomChange)="onZoomChange($event)"
      [showToolbar]="true"
      [showSidebarButton]="true"
      [showFindButton]="true"
      [showPagingButtons]="true"
      [showDrawEditor]="false"
      [showStampEditor]="false"
      [showTextEditor]="false"
      [showZoomButtons]="true"
      [showPresentationModeButton]="true"
      [showOpenFileButton]="false"
      [showPrintButton]="false"
      [showDownloadButton]="true"
      [showSecondaryToolbarButton]="true"
      [showRotateButton]="false"
      [showHandToolButton]="false"
      [showScrollingButton]="true"
      [showSpreadButton]="false"
      [showPropertiesButton]="true"
      [sidebarVisible]="false"
      [height]="'85%'"
      (pagesLoaded)="onPagesLoaded($event)"                    
      ></ngx-extended-pdf-viewer>
 </div> 
```             

Child Component.TS (textbox form field - on focus event)
```
onFocusIn(e: any) {
    this.drawRectangleOnMainCanvas(false);
  }
```

**Comment by juqing27** [comment]
> I like that idea too. Unfortunately, for us, these PDFs could be generated (aka "printed" to PDF) or scanned in, and not have the fillable form embedded in them.
> 
> I do have bounding region coordinates from an OCR that is higher in the process where I "know" where the fields exist on a page, and I can write them directly on the canvas but the problems are:
> 
> 1. The rectangle needs to scale upon resizing of the pdf (PDF Viewer normal controls used for resizing)
> 2. The rectangle needs to be removed if another field is focused in the adjacent form  (could work if we just "cleared all rectangles")
> 3. Being that I'm writing directly to the canvas, ctx.clearRect(...) will blank out the pdf page of the canvas.
> 
> I feel I'm really close. For example, here are my two Typescript functions that I'm using to create the rectangles plus do the conversion from inches to pixels for my bounding regions.
> 
> ```ts
> // =======================================
> drawRectangleOnMainCanvas(clear:boolean = false): void {
>     if (this.boundingRegion && this.boundingRegion.trim() !== '') {
>       try {
>         // Parse the string as JSON
>         const data = JSON.parse(this.boundingRegion);
>         if (data && data.length > 0) {
>           const points = data[0][0].polygon; // Take the first polygon points
>     
>           const canvasWidth = 507;
>           const canvasHeight = 657;
>     
>           // Make sure canvas element is selected after it is available in the DOM
>           const canvas = document.querySelector('canvas') as HTMLCanvasElement;
>     
>           if (canvas) {
>             this.drawSemiTransparentRectangle(canvas, points, canvasWidth, canvasHeight, clear);
>           }
>         }    
>       } catch (error) {
>         // Log the error and/or handle it as needed
>         console.error('Failed to parse bounding region data or was empty:', error);
>       }
>     }
>     
> }
> 
> drawSemiTransparentRectangle(canvas: HTMLCanvasElement, points: Array<{ x: number, y: number }>, pageWidth: number, pageHeight: number, clear: boolean = false): void {
>   if (!canvas.getContext) {
>       return;
>   }
>   
>   const ctx = canvas.getContext('2d');
>   if (!ctx) {
>       return;
>   }
> 
>   const dpi = 96; // Standard screen DPI
> 
>     // Convert inches to pixels
>     const xs = points.map(p => p.x * dpi);
>     const ys = points.map(p => p.y * dpi);
>     
>     let minX = Math.min(...xs);
>     let minY = Math.min(...ys);
>     let maxX = Math.max(...xs);
>     let maxY = Math.max(...ys);
> 
>     // Adjust for negative width and height by setting starting point to top-left corner
>     const width = maxX - minX;
>     const height = maxY - minY;
>     if (width < 0) {
>         minX += width;
>     }
>     if (height < 0) {
>         minY += height;
>     }
> 
>     //ctx.clearRect(0, 0, canvas.width, canvas.height); // Clear previous drawing
>     ctx.scale(this.currentPdfScale, this.currentPdfScale); // Apply the scale
>     if (clear) {
>       //ctx.fillStyle = 'rgba(0, 0, 255, 0.0)';
>       ctx.clearRect(0, 0, canvas.width, canvas.height); // Clear previous drawing
>     } else {
>       ctx.fillStyle = 'rgba(0, 0, 255, 0.5)';
>     }
>     
>     ctx.fillRect(minX, minY, Math.abs(width), Math.abs(height)); // Use absolute values for width and height
> }
> // =======================================
> ```
> 
> Additionally, I tried to install the annotatepdf into my angular project and it was complaining about 'fs' being used by annotatepdf. Apparently, isn't not happy about that, so I have abandoned that route for the time being.

@jcurtisKSM On the "complain about fs being used..." note, have you set "fs" to false as directed on highKite/pdfAnnotate page? See snapshot below. I did that and don't seem to have that issue for now.
![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/34175164/f895b13d-c235-4cb4-a176-faf7aad1c0f9)

**Reply by jcurtisKSM** [reply]
@juqing27 - That was the issue for annotatePDF setup, thank you.  Glossing over the readme is a thing. :)

**Comment by jcurtisKSM** [comment]
@juqing27 Circling back to this.  Do you have some more code to point to on using the this.annotationFactory = new AnnotationFactory(data);?  I have my Angular project setup so that I can instantiate PDFViewApplication  by the call below.

`const pdfViewerApplication = (window as any).PDFViewerApplication;`

The problem is that it gets loaded, but can't do much more because the "pdfDocument" property is null, so the getData() will not work and this seems to be the gateway to the AnnotationFactory needed to create our rectangles.

```ts
pdfViewerApplication.pdfDocument.getData().then((data: Uint8Array) => {
          this.annotationFactory = new AnnotationFactory(data);
          // Do something with the PDF data...
        });
```
In order to get the global varialbe for PDFViewApplication from pdf.js, I created a file [src/typings/pdf-viewer-application.d.ts] with the following code:
```ts
declare global {
    const PDFViewerApplication: {
      pdfDocument: {
        getData(): Promise<Uint8Array>;
      };
      // Add other members of PDFViewerApplication that you are using here...
    };
  }
  
  export {};
```
And then I added this to the tsconfig.json:
```json
"typeRoots": [
      "node_modules/@types",
      "src/typings"
    ],
```
The other struggle I have besides the setup here is that my PDF viewer exists in the parent component html, and fields that need to reference it exist in the parent component and two child components.
```
ParentComponent.html / TS
-- PDF Viewer
-- Fields
---- ChildComponent1.html / TS
------ Fields
---- ChildComponent2.html / TS
------ Fields
```
I appreciate both of your help on this as it's a mixture of not knowing Angular in and out combined with creating something very interactive.

**Reply by juqing27** [reply]
Hi @jcurtisKSM, same here - I'm pretty new to Angular and trying to actually do something with it :D... Glad if I could be of any help.

I haven't had time reading your code thoroughly, but wonder if you have seen my other post #2134  (mentioned in this thread too). I was asking about a issue regarding locating things on pdf, but the code is complete (with both .html and .ts files) and generally working. You can copy and try that on your end as a start I suppose.

I suspect I ran into similar issue and that's why I added the `onPdfLoaded()` function to verify the status of things. Also, for the global object, I simply declare it like this at the top of the component:
`declare var PDFViewerApplication: any;`
I don't know if this is the best way, but it works for me for now.

**Reply by stephanrauh** [reply]
As for typing: both approaches are clever. `any` is frowned upon by enterprise programmers, but it's popular in the greater (and anarchistic) JavaScript community. Basically, using `any` means the compiler and the IDE can't help you, and the more the application grows, the more painful this gets. 

As for `pdfViewerApplication.pdfDocument` being `null`: it stops being `null` as soon as the PDF document has been loaded. I guess the `onPdfLoaded()` hint is an adequate solution.

Regarding your component tree: I don't know what the problem is. I can only guess. Is the problem that the components can't communicate with each other? If so, Angular has several solutions:

- Services
- NgRX/Store (but only if your application is huge or if you've already experience with NgRX/Store)
- Signals
- Observables (but that's usually a pain in the arse, even if everybody and the grandma tells you it's the industry standard)
- Passing parameters via `@Input()` and `@Output` (this is cumbersome because you need to communicate via the common parent object, and it results in a lot of boiler-plate code passing data to and fro)
- `@ViewChild()` / `@ViewChildren()´ (this only works if your fields and the PDF viewer have the same parent component, without nesting) (and it leads to close coupling of the component, which is detrimental in the long run)
- using `document.querySelector()` or `document.getElementById()` (do that only if you're really desperate - Angular believes it owns the DOM, so it gets confused if you start messing with it)

**Comment by jcurtisKSM** [comment]
@stephanrauh 

I have reviewed the other posts and I think that for my project, I've got it down to the zoom factor calculations.  All I'm looking for is the zoom percentage.  Obviously, if it's 100% (or 1), that's easy.  I am wondering and hoping that this is a property that can be referenced in case it's like "Automatic Zoom" or something, so we could see like a 113% (or 1.13) to use for calculations.  I'm using a default dpi of 96, which my rectangles show up correctly at 100% scale.

**Reply by stephanrauh** [reply]
Sure. You're looking for `(currentZoomFactor)`. I'm showing it at https://pdfviewer.net/extended-pdf-viewer/zoom:

![image](https://github.com/stephanrauh/ngx-extended-pdf-viewer/assets/3045767/cb7b3e4f-673f-4d7c-a046-9b4bbab79aaa)

https://github.com/stephanrauh/extended-pdf-viewer-showcase/blob/cd2abf56daeaf80e2fa8c028e8ef1cc4dd0335ba/src/app/extended-pdf-viewer/zoom/zoom.component.html#L21

https://github.com/stephanrauh/extended-pdf-viewer-showcase/blob/cd2abf56daeaf80e2fa8c028e8ef1cc4dd0335ba/src/app/extended-pdf-viewer/zoom/zoom.component.html#L75

https://github.com/stephanrauh/extended-pdf-viewer-showcase/blob/cd2abf56daeaf80e2fa8c028e8ef1cc4dd0335ba/src/app/extended-pdf-viewer/zoom/zoom.component.ts#L90-L92

**Reply by jcurtisKSM** [reply]
Perfect!  I can use that to multiply by my screen dpi (96) and it puts the rectangle in the correct spot now regardless of the zoom.

Last question, is there a programmatic way to reload the PDF with the currently selected "width" like Page-Fit, 100%, etc.?  Essentially, it's as if they selected one of the options for zoom in the drop down.  This refreshes the PDF view and also clears the canvas.  Seems like an easy way to reset the canvas to clear anything on it.

Good stuff @stephanrauh .

**Comment by Hely0n** [comment]
Isn't it somehow possible to draw a simple semitransparent rectangle with the .addEditorAnnotation(
InkEditorAnnotation) method? I tried to only pass rect but it seems like the path property is mandatory.

**Reply by stephanrauh** [reply]
The ink editor allows you to add a free-hand drawing. It's a bit difficult to add a filled rectangle with a fine pencil, so I guess the answer is "no". If you need filled geometric forms, please open a ticket at https://github.com/mozilla/pdf.js/issues and try to make GitHub user @ Calixteman curious (or provide a pull request). I suppose the PDF format supports solid geometric figures, it's just that there's no editor yet.

**Reply by Hely0n** [reply]
Okay, now I understand. Thank you!

**Comment by jcurtisKSM** [comment, marked as answer]
Disregard my previous comment.  I have it working now actually just drawing on the canvas itself.  I've got save/restore canvas functions that can return the canvas to it's state before I drew a rectangle on it.  The zoom is provided via the (currentZoomFactor) that @stephanrauh provided earlier.  My code also is working with Parent => Form Fields and Parent => Child Component => Form Fields, which added to the complexity in Angular TS.

Here is my code I'm using that is working for me that just draws a rectangle on the canvas.  It does not use the PDF Annotate layer, etc. as I'm just needing to show one rectangle at a time,  so I have the luxury of just "resetting" the canvas to the previous state.

```ts
drawRectangleOnMainCanvas(clear: boolean = false): void {
    if (this.boundingRegion && this.boundingRegion.trim() !== '') {
      try {
        // Parse the string as JSON
        const data = JSON.parse(this.boundingRegion);
        if (data && data.length > 0) {
          const points = data[0][0].polygon; // Take the first polygon points
          const pageNumber = data[0][0].pageNumber;

          // Make sure canvas element is selected after it is available in the DOM
          const canvas = document.querySelector('canvas') as HTMLCanvasElement;

          if (canvas) {
            this.drawSemiTransparentRectangle(canvas, points, pageNumber, clear);
          }
        }
      } catch (error) {
        // Log the error and/or handle it as needed
        console.error('Failed to parse bounding region data or was empty:', error);
      }
    }
  }

  // Call this method to save the current state of the canvas
  saveCanvasState(canvas: HTMLCanvasElement) {
    const context = canvas.getContext('2d');
    if (context) {
      this.originalCanvasImageData = context.getImageData(0, 0, canvas.width, canvas.height);
    }
  }

  // Call this method to restore the saved state of the canvas
  restoreCanvasState(canvas: HTMLCanvasElement) {
    const context = canvas.getContext('2d');
    if (this.originalCanvasImageData && context) {
      context.putImageData(this.originalCanvasImageData, 0, 0);
    }
  }

  drawSemiTransparentRectangle(canvas: HTMLCanvasElement, points: Array<{ x: number, y: number }>, pageNumber: number, clear: boolean = false): void {
    const scale = this.zoom;

    if (!canvas.getContext) {
      return;
    }

    const ctx = canvas.getContext('2d');
    if (!ctx) {
      return;
    }

    const dpi = 96; // Standard screen DPI
    const scaledDpi = dpi * scale;

    // Convert inches to pixels
    const xs = points.map(p => p.x * scaledDpi);
    const ys = points.map(p => p.y * scaledDpi);

    let minX = Math.min(...xs);
    let minY = Math.min(...ys);
    let maxX = Math.max(...xs);
    let maxY = Math.max(...ys);

    // Adjust for negative width and height by setting starting point to top-left corner
    const width = maxX - minX;
    const height = maxY - minY;
    if (width < 0) {
      minX += width;
    }
    if (height < 0) {
      minY += height;
    }

    if (clear) {
      this.restoreCanvasState(canvas);
    } else {
      this.saveCanvasState(canvas);
      ctx.fillStyle = 'rgba(0, 0, 255, 0.5)';
      ctx.fillRect(minX, minY, Math.abs(width), Math.abs(height)); // Use absolute values for width and height
    }
  }
```
Thanks again to @stephanrauh and @juqing27 for helping out.

**Reply by stephanrauh** [reply]
@jcurtisKSM Do you think I should collect your insight and make a showcase demo for it?

**Reply by jcurtisKSM** [reply]
If you want to do that, I'm ok with it.  I know it's a different take than point-click drawing, but for anyone that wants a rectangle, and that can feed it data points, it seems helpful.   

I also added to the code to "jump" to the pageNumber that is provided by the data points, but that should be easy for people to figure out.  I know that you can get fancy with the "scrolling", but hopping to the page was good enough.
